### PR TITLE
Fix the double __root__ issue for glb

### PR DESCRIPTION
**Pull request by Bloadrick**
Fix for the the double root issue: https://github.com/BabylonJS/Babylon.js/issues/6349

### CODE REVIEW

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 1]
```diff
@@ -453,21 +454,76 @@ export class GLTFLoader implements IGLTFLoader {
...
 
-        const rootNode: INode = {
-            _babylonTransformNode: this._rootBabylonMesh,
-            index: -1
-        };
+        var indexScene = this._gltf.scene || 0;
+        var scene = this._gltf.scenes[indexScene];
+
+        // Root node must be unique
+        if (scene.nodes.length != 1) {
```
Comment (deltakosh) [inline loaders/src/glTF/2.0/glTFLoader.ts]:
```suggestion
        if (scene.nodes.length !== 1) {
```

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 2]
```diff
@@ -453,21 +454,76 @@ export class GLTFLoader implements IGLTFLoader {
...
+            return null;
+        }
+        else if(this._gltf.nodes != null) {
```
Comment (deltakosh) [inline loaders/src/glTF/2.0/glTFLoader.ts]:
```suggestion
        else if(this._gltf.nodes !== null) {
```

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 3]
```diff
@@ -453,21 +454,76 @@ export class GLTFLoader implements IGLTFLoader {
...
+            var indexRoot = scene.nodes[0];
+            var root = this._gltf.nodes[indexRoot];
+
+            // Check rotation and scale
+            if(root.rotation != undefined && root.scale != undefined){
```
Comment (deltakosh) [inline loaders/src/glTF/2.0/glTFLoader.ts]: ok...stopping here but please fix all triple equals ;)

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 4]
```diff
@@ -453,21 +454,76 @@ export class GLTFLoader implements IGLTFLoader {
...
+                if (root.rotation.join(',') != "0,0,0,1" || root.scale.join(',') != "1,1,-1") {
+                    return null;
+                }
+            }
+
+            // Check name
+            if (root.name != "__root__") {
```
Comment (deltakosh) [inline loaders/src/glTF/2.0/glTFLoader.ts]: not sure we want to check this one. As long as we have the correct transforms we are fine

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 5]
```diff
@@ -453,21 +454,76 @@ export class GLTFLoader implements IGLTFLoader {
...
+                return null;
+            }
+        }else{
```
Comment (deltakosh) [inline loaders/src/glTF/2.0/glTFLoader.ts]: spaces

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 6]
```diff
@@ -453,21 +454,76 @@ export class GLTFLoader implements IGLTFLoader {
...
+            return null;
+        }
+
+        return root;
+    }
+
+    private _createRootNode(nodes : INode[]): INode {
+
+        var ExistingRootNode = this._checkForRootNode();
```
Comment (deltakosh) [inline loaders/src/glTF/2.0/glTFLoader.ts]: no capital first letter

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 7]
```diff
@@ -453,21 +454,76 @@ export class GLTFLoader implements IGLTFLoader {
...
+        var RootAlreadyImplemented = false;
+        var INodeToReturn;
```
Comment (deltakosh) [inline loaders/src/glTF/2.0/glTFLoader.ts]: need a type

**Comment by bghgary** [comment]
This change is not in the right place. The double root should be removed on export and not on import.

**Comment by deltakosh** [comment]
Ok I was under the same impression

Closing this PR then

**Comment by Bloadrick** [comment]
Hi, I think we are both right here after thinking about it again.

So 3ds max and maya are right handed, they export in Babylon which is left handed, no root here, we are changing the geometry and the transform.
Then if we export in GLTF we are doing the same operation between Babylon and GLTF.

Current state:

After that if we put a GLTF in the sandbox we need to add a root to put it in left handed coordinates.
For the moment when exporting to GLB we keep the root added previously without adding an other one.
Then if we put this glb in the sandbox, it adds another root which is not needed (that's where I got confused for this issue).

What to do:

When we export in GLB from the sandbox we need to delete that root to keep it right handed. 
I think what I did is still usefull for legacy purpose, if a glb has been exported with multiple root it would accept it and format it in the right coordinates.

If you are ok with that I can add the delete of the root node on export and keep this PR with the changes asked implemented.
It would prevent the multiple roots and keep it to one when a GLTF/GLB is loaded.

**Comment by deltakosh** [comment]
I agree we should not add a root node if not necessary. But I struggle to convince myself that we will have the cae. When will this happen if we fix the serializer?

**Comment by Bloadrick** [comment]
If a GLB has already been exported, in my test files some of them have up to 4 or 5 root nodes.
I was thinking about someone who already used it and have already a root node in his GLB if not checked in the loader it is going to add a second one.

**Comment by deltakosh** [comment]
Ok that makes to me
@bghgary  ?

**Comment by bghgary** [comment]
I don't think the loader should do anything different.

The loader must do something in order to convert the data from right-handed (glTF) to left-handed (Babylon assuming useRightHandedSystem is false). I don't think the loader should muck with the hierarchy of the glTF itself.

If the existing glTF already has a bunch of unnecessary root nodes, then it should be up to the author of the glTF to fix the problem. The main fix is to make sure the exporter does not perpetuate the issue.

**Comment by deltakosh** [comment]
Yeah you are right. We always have to switch to left...so yeah we only need the serializer part :D

**Pull request closed without merging**